test/test.impersonated.ts

```diff
@@ -0,0 +1,305 @@
+      }),
+      createGTokenMock({
+        access_token: 'abc456',
+      }),
+      nock('https://iamcredentials.googleapis.com')
+        .post(
+          '/v1/projects/-/serviceAccounts/target@project.iam.gserviceaccount.com:generateAccessToken',
+          () => {
+            return true;
+          }
+        )
+        .reply(200, {
+          accessToken: 'qwerty345',
+          expiryDate: tomorrow.toISOString(),
+        }),
+      nock('https://iamcredentials.googleapis.com')
+        .post(
+          '/v1/projects/-/serviceAccounts/target@project.iam.gserviceaccount.com:generateAccessToken',
+          () => {
+            return true;
+          }
+        )
+        .reply(200, {
+          accessToken: 'qwerty456',
+          expiryDate: tomorrow.toISOString(),
+        }),
+    ];
+    const jwt = new JWT(
+      'foo@serviceaccount.com',
+      PEM_PATH,
+      undefined,
+      ['http://bar', 'http://foo'],
+      'bar@subjectaccount.com'
+    );
+    const impersonated = new Impersonated({
+      sourceClient: jwt,
+      targetPrincipal: 'target@project.iam.gserviceaccount.com',
+      lifetime: 30,
+      delegates: [],
+      targetScopes: ['https://www.googleapis.com/auth/cloud-platform'],
+    });
+    await impersonated.request({url});
+    // Force both the wrapped and impersonated client to appear to have
+    // expired:
+    jwt.credentials.expiry_date = Date.now();
+    impersonated.credentials.expiry_date = Date.now();
+    await impersonated.request({url});
+    assert.strictEqual(impersonated.credentials.access_token, 'qwerty456');
+    scopes.forEach(s => s.done());
+  });
+  it('throws meaningful error when context available', async () => {
+    const tomorrow = new Date();
+    tomorrow.setDate(tomorrow.getDate() + 1);
+    const scopes = [
+      createGTokenMock({
+        access_token: 'abc123',
+      }),
+      nock('https://iamcredentials.googleapis.com')
+        .post(
+          '/v1/projects/-/serviceAccounts/target@project.iam.gserviceaccount.com:generateAccessToken'
+        )
+        .reply(404, {
+          error: {
+            code: 404,
+            message: 'Requested entity was not found.',
+            status: 'NOT_FOUND',
+          },
+        }),
+    ];
+    const jwt = new JWT(
+      'foo@serviceaccount.com',
+      PEM_PATH,
+      undefined,
+      ['http://bar', 'http://foo'],
+      'bar@subjectaccount.com'
+    );
+    const impersonated = new Impersonated({
+      sourceClient: jwt,
+      targetPrincipal: 'target@project.iam.gserviceaccount.com',
+      lifetime: 30,
+      delegates: [],
+      targetScopes: ['https://www.googleapis.com/auth/cloud-platform'],
+    });
+    impersonated.credentials.access_token = 'initial-access-token';
+    impersonated.credentials.expiry_date = Date.now() - 10000;
+    await assert.rejects(impersonated.request({url}), /NOT_FOUND/);
+    scopes.forEach(s => s.done());
+  });
+  it('handles errors without context', async () => {
```

Comment:
Should you also add one more test when the underlying `sourceClient` fails to generate a token?